### ISSUE TITLE
Error on Shiny exit

### ISSUE THREAD

**Issue by jcheng5** [shikokuchuo/mirai#99]
If you start and view the following Shiny app, then stop the app before 5 seconds elapses, you get an error at the console:

```
Warning message:
In max(which(!hideable)) : no non-missing arguments to max; returning -Inf

Warning: Error in : 
Warning: Error in rep_len: invalid 'length.out' value
  [No stack trace available]
From earlier call:
  4: runApp
  3: print.shiny.appobj
  1: source
```

```r
library(shiny)
library(mirai)
library(promises)

ui <- fluidPage(
  textOutput("out")
)

server <- function(input, output, session) {
  output$out <- renderText({
    mirai({
      Sys.sleep(5)
      "hello"
    }) |> promises::as.promise()
  })
}

shinyApp(ui, server)
```

**Comment by jcheng5** [comment]
OK, so everything besides `Warning: Error in :` comes from Shiny trying to simplify the stack trace on the error, which appears to not have a stack trace, maybe? If you add `options(shiny.fullstacktrace=TRUE)` to the top of the app, which turns off stack trace simplification, you only get `Warning: Error in :`.

**Comment by shikokuchuo** [comment]
Yes I can reproduce. The `Warning: Error in :` doesn't seem to be from `mirai` though so I will have to investigate more closely.

**Comment by shikokuchuo** [comment]
@jcheng5 this is resolved by PR #100.

This turns out to be an "Unhandled promise error".

If I wrap the promise upon creation with `promises::then(onRejected = unlist)` then the actual error string is returned to the text output. I've tested for example when the mirai is asked to evaluate a non-existent function.

```r
mirai({
      Sys.sleep(5)
      elephant()
    }) |> promises::as.promise()
```

I am aware that this fix may have other unintended consequences, so any guidance is much appreciated.

**Comment by shikokuchuo** [comment]
OK, I see now that Shiny is expecting the error to be returned as a 'simpleError' condition, which is then rejected.

I now have this working for the case when `options(shiny.fullstacktrace=TRUE)` is set and the error message is shown in red in the UI. But when it isn't then I get similar warnings to the above in the console instead. I'll have to figure out how to provide the full stack trace I guess.

**Comment by shikokuchuo** [comment]
@jcheng5 the error you encountered upon closing the app is now fixed by [67ad356](https://github.com/shikokuchuo/mirai/commit/67ad3569c77200a6c2f05c4b16b43292d479dcd7).

I hope you can confirm this works for you.

I'll investigate further on getting the full stack trace for general errors returned from the mirai themselves - if you have any pointers do let me know. Thanks.

**Comment by shikokuchuo** [comment]
@jcheng5 this is now completely fixed. Through a further combination of aea530c and 1be432f.

The mirai error condition that is returned needed to be re-thrown and caught in a `tryCatch()` in the promise action for the stack trace to register.

However, even more esoteric than this was that this simply didn't work as the `as.promise.mirai` method was defined in `mirai`'s package namespace, as opposed to the 'Future' method, which is defined in `promises` itself. I verified this by inserting the mirai method into a local copy of 'promises'. 

The good news is, I did find a solution in the end that doesn't require any changes on your side, which just involved swapping the method's environment on package load. You're welcome to test the latest development build.

Also cc. @wlandau FYI

**Comment by shikokuchuo** [comment]
I realise that as I have referred directly to the 'promises' codebase to resolve this issue, I have added the relevant copyright notice in the code, and also an upfront attribution in the readme, in [370e5e6](https://github.com/shikokuchuo/mirai/commit/370e5e657b63e4711c0c9f9766af5d245b01c7f2)  @jcheng5

**Comment by jcheng5** [comment]
> I'll investigate further on getting the full stack trace for general errors returned from the mirai themselves - if you have any pointers do let me know.

The main trick here, is to know that 1) R does not put any stack trace information on the error object itself, neither when it's created nor when it's thrown; and 2) in a try/catch error handler, it's already too late to gather this information. So what you need to do is use a `withCallingHandler` which is called back at the moment an error is thrown (as opposed to `tryCatch` which is called only when the call stack has unwound to where the `tryCatch` is), grab a snapshot of the stack, and put it as an attribute on the error.

[Here's a link](https://github.com/rstudio/shiny/commit/d9a016f94cd146583e20df998854de214ed05553#diff-6449ac7bd5043d84a3399b400b25f3ea4b85a061b7d4bbfc0e73d5b77f9d4a34R25-R56) to a very early version of Shiny's logic for this. (The version on our main branch is much, much more complicated because it's designed to work well in async and also works very hard to simplify the stack trace--not super successfully I might add.) You'd need to do something similar to this around Mirai's daemon's eval call.

BTW, I suspect rlang leapfrogged ahead of Shiny in its capturing/handling of stack traces, with the exception of async I'm assuming. https://rlang.r-lib.org/reference/global_entrace.html

**Comment by jcheng5** [comment]
> However, even more esoteric than this was that this simply didn't work as the as.promise.mirai method was defined in mirai's package namespace, as opposed to the 'Future' method, which is defined in promises itself. I verified this by inserting the mirai method into a local copy of 'promises'.

I don't think you need to do any of this custom S3 registration stuff if you are OK with only supporting R >=3.6.0. You can replace `#' @method as.promise mirai` with `#' @exportS3Method promises::as.promise` and that should be enough.

https://github.com/shikokuchuo/mirai/pull/103

**Comment by jcheng5** [comment]
Here's a second PR that builds on the previous one (it doesn't work when I target it against your main branch; the main branch's overwriting of `as.promise.mirai`'s environment really messes with its lexical scope). This one should make the error stack trace look mostly correct.

https://github.com/jcheng5/mirai/pull/1

**Comment by shikokuchuo** [comment]
Thank you so much @jcheng5 for all of these. Both PRs have now been merged and deep stack traces in Shiny seem to work correctly for me.

I will also investigate your suggestion to use `withCallingHandlers()` to capture stack information from within the mirai itself. I think that this shouldn't be a problem in itself as it is only on the error path that all this additional processing needs to take place, nevertheless I would like to see if there is a way to keep the 'miraiError' as simple as possible.

**Comment by shikokuchuo** [comment]
The mirai error stack trace is now implemented in [2723789](https://github.com/shikokuchuo/mirai/commit/27237892787c378fdebeeba6e93e6c9f04b60547) by a combination of calling handlers and restarts. I've also implemented a custom $ method for 'miraiError' objects for accessing it easily.

It eliminates the calls created by the implementation itself, which should hopefully make it less confusing and more useful. E.g. see the third example here: https://shikokuchuo.net/mirai/articles/mirai.html#errors-interrupts-and-timeouts

@jcheng5 I believe this rounds out all the points raised on this issue, may I now consider this closed? Thanks again for all  the ideas and assistance!
